rat/variables.py

```diff
@@ -107,36 +105,40 @@ def get_numpy_indices(self, df):
             "__index"
         ].to_numpy(dtype=int)
 
+    def get_first_in_group_indicators(self, shifts):
```

Comment:
Can you explain this code bit?

Reply:
Yeah, for any recursive assignment where there is more than one subscript, the non-shifted subscripts will be grouping variables.

When scanning, each group should be computed independently of each other. To achieve this we reset the carry when we index into a new group (other option would be a for loop over all groups, or a vmap across all groups -- the first would be slow to compile and I'm not quite sure how to implement the second though I'm interested). Look at the `jax.lax.cond` statement in the following code. This works because the groups are sorted (if all the different groups were jumbled together, the reset wouldn't work):

```python
def transform_parameters(data, subscripts, first_in_group_indicators, parameters):
    # skills
    def scan_function_7(carry, index):
        carry = jax.lax.cond(first_in_group_indicators['skills'][index], lambda : (0.0), lambda : carry)
        (carry0) = carry
        next_value = carry0 + parameters['epsilon'][subscripts['team_year__4'][index]]
        return (next_value), next_value

    _, parameters['skills'] = jax.lax.scan(scan_function_7, (0.0), jax.numpy.arange(15))

    return parameters
```